### PR TITLE
Deprecate save_function from model checkpoint callback

### LINKED ISSUE #3875: Move the save function to the checkpoint callback
the save function is separated but it's hard to read

### PR DESCRIPTION
**Pull request by ananthsub**
## What does this PR do?

Fixes #3875

This is minor cleanup to avoid the indirection from the trainer and checkpoint callback. Enforcing that saving goes through the trainer will avoid potential bugs for items like deepspeed or sharded model saving. 

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [ ] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Review by Borda — APPROVED** [review]

**Comment by carmocca** [comment]
Should we deprecate this first? Override the `save_function` setter and show a warning

File: pytorch_lightning/callbacks/model_checkpoint.py [hunk 1]
```diff
@@ -255,8 +253,7 @@ def on_load_checkpoint(self, callback_state: Dict[str, Any]):
     def save_checkpoint(self, trainer, unused: Optional = None):
         """
         Performs the main logic around saving a checkpoint.
-        This method runs on all ranks, it is the responsibility of `self.save_function`
-        to handle correct behaviour in distributed training, i.e., saving only on rank 0.
+        This method runs on all ranks.
```
Comment (awaelchli) [inline pytorch_lightning/callbacks/model_checkpoint.py:256]: I would not remove this information. It should be good to know that it's not the callback that does the actual save operation.

File: pytorch_lightning/callbacks/model_checkpoint.py [hunk 2]
```diff
@@ -213,7 +212,6 @@ def on_pretrain_routine_start(self, trainer, pl_module):
         When pretrain routine starts we build the ckpt dir on the fly
         """
         self.__resolve_ckpt_dir(trainer)
-        self.save_function = trainer.save_checkpoint
```
Comment (awaelchli) [inline pytorch_lightning/callbacks/model_checkpoint.py:216]: couldn't find any references in our callbacks or bolts, but perhaps leave it for backward compatibility?

**Comment by ananthsub** [comment]
I'll mark this as deprecated on the callback instead of removing it outright in this PR

**Review by awaelchli — APPROVED** [review]

**Review by tchaton — APPROVED** [review]
LGTM

**Pull request merged**